### PR TITLE
feat(roll): roll Playwright 1.15.0-next-1631091227000

### PR DESCRIPTION
**Pull request by mxschmitt**
https://github.com/microsoft/playwright/pull/8632

### CODE REVIEW

File: playwright/_impl/_element_handle.py [hunk 1]
```diff
@@ -343,9 +342,25 @@ async def wait_for_selector(
         state: Literal["attached", "detached", "hidden", "visible"] = None,
         timeout: float = None,
         strict: bool = None,
-    ) -> Optional["ElementHandle"]:
-        return from_nullable_channel(
-            await self._channel.send("waitForSelector", locals_to_params(locals()))
+    ) -> "ElementHandle":
+        return cast(
+            ElementHandle,
+            from_nullable_channel(
```
Comment (kumaraditya303) [inline playwright/_impl/_element_handle.py:348]:
```suggestion
            from_channel(
```

Also `cast` is redundant here

File: playwright/_impl/_frame.py [hunk 2]
```diff
@@ -276,9 +276,25 @@ async def wait_for_selector(
         strict: bool = None,
         timeout: float = None,
         state: Literal["attached", "detached", "hidden", "visible"] = None,
-    ) -> Optional[ElementHandle]:
-        return from_nullable_channel(
-            await self._channel.send("waitForSelector", locals_to_params(locals()))
+    ) -> ElementHandle:
+        return cast(
+            ElementHandle,
+            from_nullable_channel(
```
Comment (kumaraditya303) [inline playwright/_impl/_frame.py:282]:
```suggestion
            from_channel(
```

No need for `cast` here

File: playwright/_impl/_network.py [hunk 3]
```diff
@@ -148,6 +145,29 @@ def failure(self) -> Optional[str]:
...
+            )
+        )
+
+    async def _get_headers_if_needed(self) -> List[Header]:
+        if not self._all_headers_future:
+            self._all_headers_future = asyncio.Future()
+            response = await self.response()
+            if not response:
+                return self._headers
+            headers = cast(
```
Comment (kumaraditya303) [inline playwright/_impl/_network.py:165]: I think you can just cast without an explicit `cast`

File: playwright/_impl/_tracing.py [hunk 4]
```diff
@@ -35,14 +35,35 @@ async def start(
     ) -> None:
         params = locals_to_params(locals())
         await self._channel.send("tracingStart", params)
+        await self._channel.send("tracingStartChunk")
+
+    async def start_chunk(self) -> None:
+        params = locals_to_params(locals())
```
Comment (kumaraditya303) [inline playwright/_impl/_tracing.py:41]: No params here

File: playwright/_impl/_tracing.py [hunk 5]
```diff
@@ -35,14 +35,35 @@ async def start(
...
+        await self._channel.send("tracingStop")
+
+    async def _do_stop_chunk(self, path: Union[pathlib.Path, str] = None) -> None:
+        artifact = cast(
+            Optional[Artifact],
+            from_nullable_channel(
+                await self._channel.send(
+                    "tracingStopChunk",
+                    {
+                        "save": True if path else False,
```
Comment (kumaraditya303) [inline playwright/_impl/_tracing.py:58]:
```suggestion
                        "save": bool(path),
```

File: tests/async/test_network.py [hunk 6]
```diff
@@ -538,25 +538,21 @@ async def test_network_events_should_fire_events_in_proper_order(page, server):
...
-    events = []
-    page.on("request", lambda request: events.append(f"{request.method} {request.url}"))
-    page.on(
-        "response", lambda response: events.append(f"{response.status} {response.url}")
-    )
-    page.on("requestfinished", lambda request: events.append(f"DONE {request.url}"))
-    page.on("requestfailed", lambda request: events.append(f"FAIL {request.url}"))
-    server.set_redirect("/foo.html", "/empty.html")
     FOO_URL = server.PREFIX + "/foo.html"
+    events = {}
```
Comment (kumaraditya303) [inline tests/async/test_network.py:542]: Better to use `defaultdict` here

File: playwright/_impl/_network.py [hunk 7]
```diff
@@ -263,16 +285,34 @@ def status_text(self) -> str:
...
-        return parse_headers(self._initializer["headers"])
+        return self._headers.copy()
+
+    async def all_headers(self) -> Dict[str, str]:
+        return headers_array_to_object(await self._get_headers_if_needed(), True)
+
+    async def headers_array(self) -> List[List[str]]:
+        return list(
+            map(
+                lambda header: [header["name"], header["value"]],
```
Comment (kumaraditya303) [inline playwright/_impl/_network.py:296]: Better to make it a tuple

File: tests/async/test_network.py [hunk 8]
```diff
@@ -230,18 +224,81 @@ def handle_something(request):
...
+                ['header-a', 'value-a'],
+                ['header-b', 'value-b'],
+                ['header-a', 'value-a-1'],
+                ['header-a', 'value-a-2'],
+            ]
+            })
+        """
+        )
+    request = await request_info.value
+    assert (await request.headers_array()).sort() == expected_headers.sort()
```
Comment (kumaraditya303) [inline tests/async/test_network.py:254]:
```suggestion
    assert sorted(await request.headers_array()) == sorted(expected_headers)
```
This is wrong as `[].sort()` returns `None`

File: tests/async/test_network.py [hunk 9]
```diff
@@ -230,18 +224,81 @@ def handle_something(request):
...
+
+
+async def test_should_report_response_headers_array(
+    page: Page, server: Server, is_win, browser_name
+) -> None:
+    if is_win and browser_name == "webkit":
```
Comment (kumaraditya303) [inline tests/async/test_network.py:260]: Use `is_webkit`

File: playwright/_impl/_network.py (hunk 7 above)
Comment (mxschmitt) [reply playwright/_impl/_network.py:296]: we want to keep it the same in all the languages.

**Pull request merged**